www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx

```diff
@@ -0,0 +1,239 @@
+});
+
+// Add a new website
+app.post("/api/website", async (c) => {
+  const db = drizzle(c.env.DB);
+  const data = await c.req.json();
+
+  const [newWebsite] = await db.insert(schema.websites)
+    .values({
+      url: data.url,
+      name: data.name,
+      checkInterval: data.checkInterval,
+      createdAt: new Date().toISOString(),
+    })
+    .returning();
+
+  // Create monitor for the new website
+  const monitor = c.env.SCHEDULED_MONITOR.get(
+    c.env.SCHEDULED_MONITOR.idFromName(newWebsite.id.toString())
+  );
+  await monitor.fetch(new Request('https://monitor/schedule?websiteId=' + newWebsite.id));
+
+  return c.json(newWebsite);
+});
+```
+## Calculating Uptime Percentages
+
+We also implement an endpoint to calculate uptime percentages for any given website.
+The uptime percentage calculation is more sophisticated than it might appear at first glance.
+Rather than just storing a running total, it calculates uptime dynamically based on historical check data.
+This allows for flexible time windows - you can calculate uptime for the last day, week, or month without needing to store separate aggregates.
+The calculation uses SQL's date handling capabilities to filter checks within the specified time period, then simply divides the number of successful checks by total checks.
+What's clever about this approach is that it automatically accounts for varying check intervals - a site checked every minute will have more data points than one checked every hour, but both will give accurate uptime percentages.
+The results are rounded to two decimal places to provide meaningful precision without false accuracy.
+
+```typescript
+app.get('/websites/:id/uptime', async (c) => {
+  const { id } = c.req.param();
+  const days = c.req.query('days') ?? 7;
+  const db = drizzle(c.env.DB);
+
+  const startDate = new Date();
+  startDate.setDate(startDate.getDate() - days);
+
+  const checks = await db.select()
+    .from(schema.uptimeChecks)
+    .where(
+      and(
+        eq(schema.uptimeChecks.websiteId, id),
+        sql`${schema.uptimeChecks.timestamp} >= ${startDate.toISOString()}`
+      )
+    );
+
+  const upCount = checks.filter(check => check.isUp).length;
+  const uptimePercentage = (upCount / checks.length) * 100;
+
+  return c.json({
+    uptimePercentage: Math.round(uptimePercentage * 100) / 100,
+    period: `${days} days`
+  });
+});
+```
+
+## The Web Interface
+
+Finally, we create a simple web interface to view monitored websites and their uptime percentages.
+The frontend of the uptime monitor is built using Hono's JSX rendering capabilities, featuring two main components: a `Layout` component and a `WebsiteList` component.
+The `Layout` component serves as a base template, providing a clean, flexible structure with CSS-in-JS styling that ensures full viewport height and proper padding.
+The `WebsiteList` component is where the main UI lives, displaying monitored websites in a responsive grid layout
+
```

Comment:
Do you have a screenshot of the UI or is this worker somewhere live and we can share a link to it